File: langs/csharp/src/VMx/Aggregates/AggregateVM1.cs
using VMx.Builders;
using VMx.Components;
using VMx.Services;

namespace VMx.Aggregates;

/// <summary>
/// Arity-1 aggregate viewmodel. A fixed tuple of one heterogeneous component VM.
/// The component slot is populated lazily on construct() by invoking the factory
/// provided via the nested builder.
///
/// See spec/08-aggregate-vm.md and ADR-0007.
/// </summary>
/// <typeparam name="VM1">Type of the first component.</typeparam>
public sealed class AggregateVM1<VM1> : ComponentVMBase, IAggregateVM1<VM1>, IAggregateSlots
    where VM1 : class, IComponentVM
{
    private readonly Func<VM1> _factory1;
    private VM1? _component1;

    IEnumerable<IComponentVM> IAggregateSlots.EnumerateSlots()
    {
        if (_component1 is { } c1) yield return c1;
    }

    // ── IAggregateVM1<VM1> ──────────────────────────────────────────────────

    /// <inheritdoc/>
    public VM1? Component1 => _component1;

    // ── IComponentVM.Type ───────────────────────────────────────────────────

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Aggregate;

    // ── Constructor ─────────────────────────────────────────────────────────

    private AggregateVM1(
        string name,
        string hint,
        IMessageHub hub,
        IDispatcher dispatcher,
        Func<VM1> factory1)
        : base(name, hint, hub, dispatcher, onConstruct: null, onDestruct: null)
    {
        _factory1 = factory1;
    }

    // ── Lifecycle overrides ─────────────────────────────────────────────────

    /// <inheritdoc/>
    protected override void OnConstruct()
    {
        // On Reconstruct, the previous slot instance is in Destructed state but
        // still holds hub subscriptions and command Subjects. Dispose it before
        // overwriting so subscribers don't leak across the Reconstruct boundary.
        _component1?.Dispose();
        _component1 = _factory1();
        NotifyPropertyChanged(nameof(Component1));

        CompleteLifecycleHookAfter(TransitionChildrenAsync(
            [_component1], construct: true));
    }

    /// <inheritdoc/>
    protected override void OnDestruct()
    {
        CompleteLifecycleHookAfter(TransitionChildrenAsync(
            new IComponentVM?[] { _component1 }.OfType<IComponentVM>(),
            construct: false));
    }

    /// <summary>
    /// Dispose cascade (LIFE-013): dispose each component slot depth-first, then self.
    /// </summary>
    public override void Dispose()
    {
        var firstError = DisposeChildren([_component1]);
        try { base.Dispose(); }
        catch (Exception error)
        {
            firstError ??= System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(error);
        }
        firstError?.Throw();
    }

    // ── Builder factory ─────────────────────────────────────────────────────

    /// <summary>Returns a new empty builder for <see cref="AggregateVM1{VM1}"/>.</summary>
    public static AggregateVM1Builder Builder() => new();

    // ── Nested builder ──────────────────────────────────────────────────────

    /// <summary>
    /// Immutable fluent builder for <see cref="AggregateVM1{VM1}"/>.
    /// Each setter returns a NEW builder instance (BLD-001).
    /// </summary>
    public sealed class AggregateVM1Builder
    {
        private readonly string? _name;
        private readonly string _hint;
        private readonly IMessageHub? _hub;
        private readonly IDispatcher? _dispatcher;
        private readonly Func<VM1>? _factory1;

        internal AggregateVM1Builder()
        {
            _hint = "";
        }

        private AggregateVM1Builder(
            string? name,
            string hint,
            IMessageHub? hub,
            IDispatcher? dispatcher,
            Func<VM1>? factory1)
        {
            _name = name;
            _hint = hint;
            _hub = hub;
            _dispatcher = dispatcher;
            _factory1 = factory1;
        }

        /// <summary>Sets the required Name field.</summary>
        public AggregateVM1Builder Name(string name)
            => new(name, _hint, _hub, _dispatcher, _factory1);

        /// <summary>Sets the optional Hint field (default: empty string).</summary>
        public AggregateVM1Builder Hint(string hint)
            => new(_name, hint, _hub, _dispatcher, _factory1);

        /// <summary>Sets the required Services (hub + dispatcher).</summary>
        public AggregateVM1Builder Services(IMessageHub hub, IDispatcher dispatcher)
            => new(_name, _hint, hub, dispatcher, _factory1);

        /// <summary>Sets the required factory for Component1.</summary>
        public AggregateVM1Builder Component1(Func<VM1> factory)
            => new(_name, _hint, _hub, _dispatcher, factory);

        /// <summary>
        /// Validates required fields and constructs a <see cref="AggregateVM1{VM1}"/>.
        /// Throws <see cref="BuilderValidationException"/> if a required field is missing.
        /// </summary>
        public AggregateVM1<VM1> Build()
        {
            BuilderValidationException.Require(_name, "Name");
            BuilderValidationException.Require(_hub, "Hub");
            BuilderValidationException.Require(_dispatcher, "Dispatcher");
            BuilderValidationException.Require(_factory1, "Component1");

            return new AggregateVM1<VM1>(_name, _hint, _hub, _dispatcher, _factory1);
        }
    }
}

File: langs/typescript/src/aggregates/aggregateVM1.ts
/**
 * AggregateVM1<VM1> — arity-1 aggregate viewmodel.
 *
 * See spec/08-aggregate-vm.md and ADR-0007.
 */
import { ComponentVMBase } from "../components/componentVMBase.js";
import { ViewModelType } from "../components/types.js";
import { PropertyChangedMessage } from "../messages/propertyChanged.js";
import type { IMessageHub } from "../services/messageHub.js";
import type { IDispatcher } from "../services/dispatcher.js";
import { BuilderValidationError } from "../builders/exceptions.js";

const SENTINEL = Symbol("not-set");

export class AggregateVM1<VM1 extends ComponentVMBase> extends ComponentVMBase {
  readonly #factory1: () => VM1;
  #component1: VM1 | null = null;

  constructor(opts: {
    name: string;
    hint: string;
    hub: IMessageHub;
    dispatcher: IDispatcher;
    factory1: () => VM1;
  }) {
    super(opts);
    this.#factory1 = opts.factory1;
  }

  get type(): ViewModelType {
    return ViewModelType.Aggregate;
  }

  get component1(): VM1 | null {
    return this.#component1;
  }

  protected override _onConstruct(): void {
    // On Reconstruct, the previous slot instance is in Destructed state but
    // still holds hub subscriptions and command Subjects. Dispose it before
    // overwriting so subscribers don't leak across the Reconstruct boundary.
    this.#component1?.dispose();
    this.#component1 = this.#factory1();
    this._hub.send(PropertyChangedMessage.create(this, this._name, "component1"));
    this._raisePropertyChanged("component1");
    this.#component1.construct();
  }

  protected override _onDestruct(): void {
    this.#component1?.destruct();
  }

  override dispose(): void {
    // Depth-first dispose (LIFE-013): each component slot first, then self.
    // Matches the override pattern used by CompositeVMBase/GroupVM so that
    // subscribers observe child Disposed transitions before the aggregate's
    // own Disposed transition — a single dispose-ordering rule across all
    // container VM kinds.
    this.#component1?.dispose();
    super.dispose();
  }

  static builder<VM1 extends ComponentVMBase>(): AggregateVM1Builder<VM1> {
    return new AggregateVM1Builder<VM1>();
  }
}

export class AggregateVM1Builder<VM1 extends ComponentVMBase> {
  #name: string | null = null;
  #hint = "";
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #factory1: (() => VM1) | typeof SENTINEL = SENTINEL;

  constructor(from?: AggregateVM1Builder<VM1>) {
    if (from) {
      this.#name = from.#name;
      this.#hint = from.#hint;
      this.#hub = from.#hub;
      this.#dispatcher = from.#dispatcher;
      this.#factory1 = from.#factory1;
    }
  }

  name(v: string): AggregateVM1Builder<VM1> { const b = new AggregateVM1Builder<VM1>(this); b.#name = v; return b; }
  hint(v: string): AggregateVM1Builder<VM1> { const b = new AggregateVM1Builder<VM1>(this); b.#hint = v; return b; }
  services(hub: IMessageHub, disp: IDispatcher): AggregateVM1Builder<VM1> { const b = new AggregateVM1Builder<VM1>(this); b.#hub = hub; b.#dispatcher = disp; return b; }
  component1(f: () => VM1): AggregateVM1Builder<VM1> { const b = new AggregateVM1Builder<VM1>(this); b.#factory1 = f; return b; }

  build(): AggregateVM1<VM1> {
    if (this.#name === null) throw new BuilderValidationError("name");
    if (this.#hub === null || this.#dispatcher === null) throw new BuilderValidationError("services");
    if (this.#factory1 === SENTINEL) throw new BuilderValidationError("component1");
    const factory1 = this.#factory1;
    return new AggregateVM1<VM1>({ name: this.#name, hint: this.#hint, hub: this.#hub, dispatcher: this.#dispatcher, factory1 });
  }
}
